### ISSUE TITLE
Cancel installing when Conan finds duplicate dependencies

### ISSUE THREAD

**Issue by monsdar** [conan-io/conan#521]
Currently there's a warning when requirements are duplicated or requirements are overriden by parts of my dependencies:

`WARN: Duplicated requirement Bar/1.0.0@monsdar/release != Bar/2.0.0@monsdar/release`
and
`Foo/1.0.0@monsdar/release requirement Bar/1.0.0@monsdar/release overriden by your conanfile to Bar/2.0.0@monsdar/release`

This usually shows that something is odd and does not fit together. Is it possible to cancel the build with an error returncode instead of just printing warnings?

**Comment by memsharded** [comment]
Those are different cases right?
The first one shows some inconsistency in diamond dependencies trees, and warns that it is keeping the first one.

The second is just telling that the downstream is overriding/redefining dependencies versions. This shouldn't be a stopper at all, it is intended to be that way, the downstream conanfile, i.e. the user one, always has precedence and can do that.

The first one is the one that indeed could cause some issues due to the inconsistency (not using ``Bar/2`` by someone who declared to use that version), but not sure that this should always throw an error. It will typically work if the version is maintained but just changed the user.

Any suggestion anyone else? What would you like the behavior to be?

**Comment by Phibedy** [comment]
I would prefer a installation flag like ````-Werror````

**Comment by memsharded** [comment]
Yes, that is probably a good idea, make it opt-in, to not break builds of someone that might be affected.

**Comment by monsdar** [comment]
Yeah, had the same in mind - opting in would be the best way to go.

I my opinion the second case could cause trouble too:
ApplicationA loads 2 plugins during startup:
* PluginA depends on LibraryA/1.0.0 (shared lib)
* PluginB depends on LibraryA/2.0.0 (shared lib)

If I would like to load both Plugins within ApplicationA I have a problem. I can just use one version of LibraryA.

I can create a simple `conanfile.txt` that loads the app and my plugins into a directory:

```
[requires]
ApplicationA/1.0.0@user/testing
PluginA/1.0.0@user/testing
PluginB/1.0.0@user/testing

[imports]
#copy all .exe and .dll into a directory
```

Which version of LibraryA is installed? One of the plugins will not work because it does not get the version of the lib it needs. In that case I would rather have `conan install` to fail directly.

Hope that makes sense.

**Comment by memsharded** [comment]
@monsdar yes, for that case, you have a problem, not strictly related to conan, but to the application components relations. I'd say it is possible to depend "privately" on some libraries, if they are embedded into the consumer artifact, i.e. Plugins A and B are shared libraries that depend on a header only or statically linked LibraryA, and the details of Library A are not exposed over the interface of the Plugins. In that case, it is possible to use different versions for LibraryA.

If LibraryA is to be a shared library, I don't know a way to make the Plugins shared libs to load a specific version of LibraryA. If several paths were added, I guess they will find the first one in the first path. I'd say that for LibraryA being a shared library, both plugins should be binary compatible with it, so if LibraryA/2.0 actually breaks the API wrt to LibraryA/1.0, indeed that layout cannot work, and one of the plugins must upgrade or downgrade.

Besides that, the second message is not allowing using both 1.0 and 2.0 versions of LibraryA, it is just informing the user that Bar/2.0.0@monsdar/release will be used instead of Bar/1.0.0@monsdar/release, for all the packages in the dependency tree. Different versions of the same package in the same dependency tree can be only achieved through "private" dependencies, but the conditions for defining them as "private" should be met. Being a shared library (as said by the name), doesn't meet condition to be a "private" dependency.

Does this make sense?

In any case, I am having a look and seeing if the behavior can be changed to error instead of warn without cluttering too much the interface or the codebase.

**Comment by memsharded** [comment]
Implemented in https://github.com/conan-io/conan/pull/536

Please note that I have applied it to actual conflicts in the dependency tree that were resolved by default:
- Duplicated dependency
- Mismatching version (or user/channel) of the same package
- Mismatching option for the same package

The case of "overriden" is OK, that message is produced when a dependency is overriden by downstream (the user), which is the desired behavior.

**Comment by monsdar** [comment]
I still think the case of overriden is not ok, as it is not always in the hand of the user and without him acknowledging it causes problems. This at least is what's our case.

```
A1 -> B -> C
A2 -> D -> C
```

The user creates C which only uses B and D. He does not know that he indirectly uses two versions of A. This is not a problem if the user is able to see the message and react on it (though it is at the very top of a potentially long build output). But it is a problem when it comes to Continuous Delivery and Deployment. How should the system know that there is a problem in the dependency tree? Parsing the output of conan?

As it is only opt-in I do not see an issue here where a user gets some behaviour he does not want to have. For CD we can then turn it on and get feedback if everything's alright.

**Comment by memsharded** [comment]
I mean, that case has already been converted to ``werror`` in the proposed PR: https://github.com/memsharded/conan/blob/a5709326456c1d011f2f3b23569a44b85feb784a/conans/client/deps_builder.py#L349

So it will throw an error if using the argument ``conan install --werror`` and abort the installation.
The message you showed in the previous comment, about ``Foo/1.0.0@monsdar/release requirement Bar/1.0.0@monsdar/release overriden by your conanfile to Bar/2.0.0@monsdar/releas`` only happens when the override is defined explicitely downstream by the user, which is a different use case.

Does this make sense? Thanks!

**Comment by monsdar** [comment]
Ah ok, then I misunderstood. :+1:

**Comment by monsdar** [comment]
I think this can be closed then?

**Comment by memsharded** [comment]
Nop :)
This feature has been implemented by: https://github.com/conan-io/conan/pull/536
It is merged to develop, it will be released in 0.14.0. So let's leave it open until released and label it as fixed :)
